lib/trino-metastore/src/main/java/io/trino/metastore/MeasuredHiveMetastore.java

```diff
@@ -0,0 +1,507 @@
+    public boolean functionExists(String databaseName, String functionName, String signatureToken)
+    {
+        return wrap("functionExists", () -> delegate.functionExists(databaseName, functionName, signatureToken));
+    }
+
+    @Override
+    public Collection<LanguageFunction> getAllFunctions(String databaseName)
+    {
+        return wrap("getAllFunctions", () -> delegate.getAllFunctions(databaseName));
+    }
+
+    @Override
+    public Collection<LanguageFunction> getFunctions(String databaseName, String functionName)
+    {
+        return wrap("getFunctions", () -> delegate.getFunctions(databaseName, functionName));
+    }
+
+    @Override
+    public void createFunction(String databaseName, String functionName, LanguageFunction function)
+    {
+        wrap("createFunction", () -> delegate.createFunction(databaseName, functionName, function));
+    }
+
+    @Override
+    public void replaceFunction(String databaseName, String functionName, LanguageFunction function)
+    {
+        wrap("replaceFunction", () -> delegate.replaceFunction(databaseName, functionName, function));
+    }
+
+    @Override
+    public void dropFunction(String databaseName, String functionName, String signatureToken)
+    {
+        wrap("dropFunction", () -> delegate.dropFunction(databaseName, functionName, signatureToken));
+    }
+
+    private void wrap(String callName, Runnable call)
+    {
+        wrap(callName, () -> {
+            call.run();
+            return null;
+        });
+    }
+
+    private <T> T wrap(String callName, Callable<T> call)
+    {
+        MetastoreApiCallStats callStats = apiCallStats.computeIfAbsent(callName, _ -> new MetastoreApiCallStats());
+        long start = ticker.read();
+        try {
+            return call.call();
+        }
+        catch (Exception e) {
+            callStats.addFailure();
+            allApiCallsStats.addFailure();
+            Throwables.throwIfUnchecked(e);
+            throw new RuntimeException(e);
+        }
+        finally {
+            long timeNanos = ticker.read() - start;
+            callStats.addTime(timeNanos);
+            allApiCallsStats.addTime(timeNanos);
+        }
+    }
+
+    @ThreadSafe
+    static class MetastoreApiCallStats
+    {
+        private final TDigest timeNanosDistribution = new TDigest();
+        private long totalTimeNanos;
+        private long totalFailures;
+
+        public synchronized void addTime(long nanos)
+        {
+            timeNanosDistribution.add(nanos);
+            totalTimeNanos += nanos;
+        }
+
+        public synchronized void addFailure()
+        {
+            totalFailures++;
+        }
+
+        public synchronized void storeTo(ImmutableMap.Builder<String, Metric<?>> metrics, String prefix)
+        {
+            // DistributionSnapshot does not retain reference to the histogram
+            metrics.put(prefix + ".time.distribution", new DistributionSnapshot(new TDigestHistogram(timeNanosDistribution)));
```

Comment:
I don't like the fact that we're recording TDigestHistograms here. We've seen before that using this on coordinator adds significantly to CPU and memory usage and also bloats queryinfo jsons. We've spent a lot of time on optimizing the existing usage of these and they're still often the topmost things in JFR profiles for coordinator.
I'm skeptical that recording a detailed distribution of time taken is even useful for us. I suggest dropping the distribution in favor of recording just total time, number of invocations and failures (e.g. io.trino.spi.eventlistener.QueryPlanOptimizerStatistics).

Reply:
Maybe keep the distribution only in `metastore.all` if you really feel convinced about the value of having it

Reply:
We are not recording `TDigestHistograms` here, but rather a `DistributionSnapshot` that only contains a limited list of percentiles. The percentiles could be useful to see problems with non-constant latency patterns.